ouroboros-network/src/Ouroboros/Network/PeerSelection/PublicRootPeers.hs

```diff
@@ -0,0 +1,208 @@
+toPublicConfigPeerSet (PublicRootPeers pp _ _ _) = Map.keysSet pp
+
+toPublicConfigPeerMap :: PublicRootPeers peeraddr -> Map peeraddr PeerAdvertise
+toPublicConfigPeerMap (PublicRootPeers pp _ _ _) = pp
+
+toBootstrapPeerSet :: PublicRootPeers peeraddr -> Set peeraddr
+toBootstrapPeerSet (PublicRootPeers _ bsp _ _) = bsp
+
+toLedgerPeerSet :: PublicRootPeers peeraddr -> Set peeraddr
+toLedgerPeerSet (PublicRootPeers _ _ lp _) = lp
+
+toBigLedgerPeerSet :: PublicRootPeers peeraddr -> Set peeraddr
+toBigLedgerPeerSet (PublicRootPeers _ _ _ blp) = blp
+
+toAllLedgerPeerSet :: Ord peeraddr => PublicRootPeers peeraddr -> Set peeraddr
+toAllLedgerPeerSet (PublicRootPeers _ _ lp blp) = lp <> blp
+
+-- | Preserves PublicRootPeers invariant. If the two sets are not disjoint,
+-- removes the common ones from the bootstrap peers set since its the most
+-- sensitive set.
+--
+fromMapAndSet :: Ord peeraddr
+              => Map peeraddr PeerAdvertise -- ^
+              -> Set peeraddr -- ^ bootstrap peers
+              -> Set peeraddr -- ^ ledger peers
+              -> Set peeraddr -- ^ big ledger peers
+              -> PublicRootPeers peeraddr
+fromMapAndSet pp bsp lp blp =
+  -- Enforcing invariants
+  let newPP = pp `Map.withoutKeys` (bsp <> lp <> blp)
+      newLP = lp \\ (Map.keysSet newPP <> bsp <> blp)
+      newBLP = blp \\ (Map.keysSet newPP <> bsp <> newLP)
+   in PublicRootPeers newPP bsp newLP newBLP
+
+insertPublicConfigPeer :: Ord peeraddr
+                       => peeraddr
+                       -> PeerAdvertise
+                       -> PublicRootPeers peeraddr
+                       -> PublicRootPeers peeraddr
+insertPublicConfigPeer p pa prp =
+  let (PublicRootPeers pp bsp lp blp) = difference prp (Set.singleton p)
+   in if Set.member p bsp
+         then prp
+         else PublicRootPeers (Map.insert p pa pp) bsp lp blp
+
+insertBootstrapPeer :: Ord peeraddr
+                    => peeraddr
+                    -> PublicRootPeers peeraddr
+                    -> PublicRootPeers peeraddr
+insertBootstrapPeer p prp =
+  let (PublicRootPeers pp bsp lp blp) = difference prp (Set.singleton p)
+   in PublicRootPeers pp (Set.insert p bsp) lp blp
+
+insertLedgerPeer :: Ord peeraddr
+                 => peeraddr
+                 -> PublicRootPeers peeraddr
+                 -> PublicRootPeers peeraddr
+insertLedgerPeer p prp =
+  let (PublicRootPeers pp bsp lp blp) = difference prp (Set.singleton p)
+   in if Set.member p bsp
+         then prp
+         else PublicRootPeers pp bsp (Set.insert p lp) blp
+
+insertBigLedgerPeer :: Ord peeraddr
+                    => peeraddr
+                    -> PublicRootPeers peeraddr
+                    -> PublicRootPeers peeraddr
+insertBigLedgerPeer p prp =
+  let (PublicRootPeers pp bsp lp blp) = difference prp (Set.singleton p)
+   in if Set.member p bsp
+         then prp
+         else PublicRootPeers pp bsp lp (Set.insert p blp)
```

Comment:
I wonder if more useful would be functions which takes a `Set peeraddr` instead of just a single `peeraddr`?

Reply:
So far, surprisingly I haven't used any of these functions, so I think we can go with the flow with these ones